### ISSUE TITLE
Session Affinity

### ISSUE THREAD

**Issue by Maxpain177** [cloudflare/cloudflare-ingress-controller#78]
Do Argo Tunnels supports Session Affinity (sticky sessions) in cluster scope?
For example, Nginx Ingress Controller allows specifying several annotations for it:
```yaml
nginx.ingress.kubernetes.io/session-cookie-name: "__lbroute"
nginx.ingress.kubernetes.io/session-cookie-hash: "sha1"
```

It's important to us, due we support old web-browsers without WebSockets, using XHR Long Polling as a fallback.

**Comment by mattalberts** [comment]
@Maxpain177 the tunnel controller doesn't expose a way to configure sticky sessions, though you are able to configure session affinity as part of the load-balancer
 - [LoadBalancer Setup](https://support.cloudflare.com/hc/en-us/articles/115000081911-Tutorial-How-to-Set-Up-Load-Balancing-Intelligent-Failover-on-Cloudflare)
 - [LoadBalancer Session Affinity Setup](https://support.cloudflare.com/hc/en-us/articles/115001771451-Load-Balancing-Session-Affinity)

**Comment by Maxpain177** [comment]
@mattalberts, Yes, I think about "Session affinity" feature as part of the CloudFlare Load Balancer. But it associates all user's requests with a single origin (i.e., Kubernetes Node in my case). I need to associate every request not only with single origin but with single Endpoint/Pod.

**Comment by mattalberts** [comment]
@Maxpain177 (sorry for the delay was digging around/investigating/fixing issues).

Correct, the current version of the ingress controller maps a "tunnel" to a service (specifically `namespace.name:port`).  I that model, you're left with the load-balancing options defined by a service (or rather the the configured proxy mode, likely ipvs or iptables).
 - https://kubernetes.io/docs/concepts/services-networking/

Have you tried setting `sessionAffinity` to `ClientIP` on your service?

**Comment by Maxpain177** [comment]
@mattalberts, Hi. ClientIP sessionAffinity does not evenly load balances when multiple clients seat behind NAT (with same IPs). It will be cool to use Endpoints API (like ingress-nginx) instead of Service and make load balancing with Cookie session affinity.

**Comment by mattalberts** [comment]
@Maxpain177 correct.  As i try to draft up changes for 0.6.x I can start looking at that.  An easy way around the restrictions (and maybe a better model than a centralized controller) would be to run `cloudflared` as a side-car container (part of your replicaset spec template used to spin up pods).  That would give you a cloudflared instance per-pod.  The solutions does come with some downsides that the controller end-arounds, but.
